public/app/features/explore/state/reducers.ts

```diff
@@ -591,12 +575,81 @@ export const itemReducer = reducerFactory<ExploreItemState>({} as ExploreItemSta
       const { loadingState } = action.payload;
       return {
         ...state,
-        loadingState,
+        queryResponse: {
+          ...state.queryResponse,
+          state: loadingState,
+        },
+        loading: loadingState === LoadingState.Loading || loadingState === LoadingState.Streaming,
       };
     },
   })
+  .addMapper({
+    //queryStreamUpdatedAction
+    filter: queryEndedAction,
+    mapper: (state, action): ExploreItemState => {
+      return processQueryResponse(state, action);
+    },
+  })
+  .addMapper({
+    filter: queryStreamUpdatedAction,
+    mapper: (state, action): ExploreItemState => {
+      return processQueryResponse(state, action);
+    },
+  })
   .create();
 
+export const processQueryResponse = (
```

Comment:
I would actually think it would make sense to have this as a selector and only process the data when needed for rendering. One optimisation I did for example is to show only 100 lines when streaming and 1000 when paused (user scrolls up) for more context. Also while paused the websocket was kept open so that we can easily resume without clearing anything and with the real most recent context.

If the processing was done only on render (with memoization for example in case multiple components need to process the same data) it would probably be performing much better as it would not process all 1000 lines but just the 100, and would not process anything when paused (but still kept the current window of logs)

Not something that needs to be changed here, but something I realised and would like to try in some separate PR.